src/streaming/downsampler.cpp

```diff
@@ -0,0 +1,295 @@
+#include "downsampler.hh"
+#include "macros.hh"
+#include "zarrv2.array.writer.hh"
+#include "zarrv3.array.writer.hh"
+
+namespace {
+template<typename T>
+[[nodiscard]] ByteVector
+scale_image(ConstByteSpan src, size_t& width, size_t& height)
+{
+    const auto bytes_of_src = src.size();
+    const auto bytes_of_frame = width * height * sizeof(T);
+
+    EXPECT(bytes_of_src >= bytes_of_frame,
+           "Expecting at least ",
+           bytes_of_frame,
+           " bytes, got ",
+           bytes_of_src);
+
+    const int downscale = 2;
+    constexpr auto bytes_of_type = static_cast<double>(sizeof(T));
+    const double factor = 0.25;
+
+    const auto w_pad = static_cast<double>(width + (width % downscale));
+    const auto h_pad = static_cast<double>(height + (height % downscale));
+
+    const auto size_downscaled =
+      static_cast<uint32_t>(w_pad * h_pad * factor * bytes_of_type);
+
+    ByteVector dst(size_downscaled, static_cast<std::byte>(0));
+    auto* dst_as_T = reinterpret_cast<T*>(dst.data());
+    auto* src_as_T = reinterpret_cast<const T*>(src.data());
+
+    size_t dst_idx = 0;
+    for (auto row = 0; row < height; row += downscale) {
+        const bool pad_height = (row == height - 1 && height != h_pad);
+
+        for (auto col = 0; col < width; col += downscale) {
+            size_t src_idx = row * width + col;
+            const bool pad_width = (col == width - 1 && width != w_pad);
+
+            auto here = static_cast<double>(src_as_T[src_idx]);
+            auto right = static_cast<double>(
+              src_as_T[src_idx + (1 - static_cast<int>(pad_width))]);
+            auto down = static_cast<double>(
+              src_as_T[src_idx + width * (1 - static_cast<int>(pad_height))]);
+            auto diag = static_cast<double>(
+              src_as_T[src_idx + width * (1 - static_cast<int>(pad_height)) +
+                       (1 - static_cast<int>(pad_width))]);
```

Comment:
I don't understand some parts of the padding/offset calculation entirely at the moment, though I'm sure this can work.

I think it's fine just to do boundary clamping inside the loop.

`row1 = max(row+1,height-1)` etc.

Alternatively, could also make a decision just to truncate the odd edges. So if you had a 3x3 input, the output is 1x1 - or if you had 513x512 the output is 256x256.

We need to document the decision and test against it since it's an important detail.

Reply:
This is one of those "too-clever-by-half" things I wrote. For example, `pad_width` is `true` iff `col` is odd and equal to `width - 1`, so if we're padding (because we're on the edge) it'll just take the same pixel twice and if we're not, it'll take the pixel adjacent to it.

Not sure what you're describing when you say boundary clamping inside the loop.